### PR TITLE
Add GraphQL client

### PR DESCRIPTION
**Pull request by SepehrGanji**

### CODE REVIEW

File: packages/graphql-client/src/graphqlClient.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            headerId
+          }
+        }
+      }
+    `;
+
+    const response = await this._client.query(query, { count });
+
+    // What is size? Is it the block size?
+    // What is `transactionsId`?
```
Comment (arobsn) [inline packages/graphql-client/src/graphqlClient.ts:116]: These are non-required properties and can be safely ignored, but `BlockHeader` is mistakenly defining it as required, can you please put a `?` in `size`, `extensionId`, `transactionsId` and `adProofsId` in `BlockHeader` type, so you can ignore them from here?

File: packages/graphql-client/src/graphqlClient.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        votes: header.votes.join(""),
+        size: 0,
+        transactionsId: "",
+        adProofsId: header.adProof.headerId,
+        extensionId: header.extension.headerId
+      })) ?? []
+    );
+  }
+  checkTransaction(transaction: SignedTransaction): Promise<boolean> {
+    // Size is missing from SignedTransaction! What to do?
```
Comment (arobsn) [inline packages/graphql-client/src/graphqlClient.ts:132]: Size is not a required field, if required by GraphQL you can temporarily set it to `0` and open an issue on `ergo-graphql` repo so we can fix that later.

File: packages/graphql-client/src/graphqlClient.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    throw new Error("Method not implemented.");
+  }
+  submitTransaction(transaction: SignedTransaction): Promise<string> {
+    // Same as checkTransaction
+    throw new Error("Method not implemented.");
+  }
+  reduceTransaction(transaction: UnsignedTransaction): Promise<string> {
+    // It's not implemented in graphql. What to do?
```
Comment (arobsn) [inline packages/graphql-client/src/graphqlClient.ts:140]: In this case you must throw `NotSupportedError`.

**Review by arobsn — CHANGES_REQUESTED** [review]

File: packages/graphql-client/package.json [hunk 4]
```diff
@@ -28,6 +28,7 @@
     "node": ">=14"
   },
   "dependencies": {
+    "@ergo-graphql/types": "^0.5.0",
```
[the remark this answers is not in the document]
Comment (SepehrGanji) [reply packages/graphql-client/package.json:31]: Done!

File: packages/graphql-client/src/graphqlClient.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+import {
+  BlockHeader,
+  BoxQuery,
+  BoxWhere,
+  ChainClientBox,
+  IChainDataClient,
+  SignedTransaction,
+  UnsignedTransaction
+} from "packages/common/src";
+import { HexString } from "sigmastate-js/main";
```
[the remark this answers is not in the document]
Comment (SepehrGanji) [reply packages/graphql-client/src/graphqlClient.ts:12]: Imported from common.

File: packages/graphql-client/src/graphqlClient.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+import { Client, createClient, fetchExchange, gql } from "@urql/core";
+import {
+  BlockHeader,
+  BoxQuery,
+  BoxWhere,
+  ChainClientBox,
+  IChainDataClient,
+  SignedTransaction,
+  UnsignedTransaction
+} from "packages/common/src";
```
[the remark this answers is not in the document]
Comment (SepehrGanji) [reply packages/graphql-client/src/graphqlClient.ts:11]: Done!

File: packages/graphql-client/src/graphqlClient.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+  constructor() {
+    this._client = createClient({
+      url: "TODO",
+      exchanges: [fetchExchange],
+      requestPolicy: "network-only"
+    });
+  }
+  async getUnspentBoxes(q: BoxQuery<ExtendedBoxWhere>): Promise<ChainClientBox[]> {
+    const query = gql<{ boxes: Omit<ChainClientBox, "confirmed">[] }, QueryBoxesArgs>`
```
[the remark this answers is not in the document]
Comment (SepehrGanji) [reply packages/graphql-client/src/graphqlClient.ts:35]: Actually when I do that I must do some modifications (like between string  and bigint) to make the code work, but this way it works without a complicated map.
Of course the actually return type of gql is `box` and I'm not sure if the gql package does a good job in casting, So probably wait for the unit tests to see if the simpler version work.
What do you think?

File: packages/graphql-client/src/graphqlClient.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+};
+
+export type ExtendedBoxQuery = BoxQuery<ExtendedBoxWhere>;
+
+export class GraphqlClient implements IChainDataClient<BoxWhere> {
+  private _client: Client;
+
+  constructor() {
+    this._client = createClient({
+      url: "TODO",
```
[the remark this answers is not in the document]
Comment (SepehrGanji) [reply packages/graphql-client/src/graphqlClient.ts:29]: Done!

File: packages/graphql-client/src/graphqlClient.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+export type ExtendedBoxWhere = BoxWhere & {
+  /** Base16-encoded BoxIds */
+  boxIds?: HexString[];
+
+  /** Base16-encoded ErgoTrees or Base58-encoded addresses */
+  contracts?: HexString[];
+};
+
+export type ExtendedBoxQuery = BoxQuery<ExtendedBoxWhere>;
```
[the remark this answers is not in the document]
Comment (SepehrGanji) [reply packages/graphql-client/src/graphqlClient.ts:22]: Done!

File: packages/graphql-client/src/graphqlClient.ts [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  BoxQuery,
+  BoxWhere,
+  ChainClientBox,
+  IChainDataClient,
+  SignedTransaction,
+  UnsignedTransaction
+} from "packages/common/src";
+import { HexString } from "sigmastate-js/main";
+
+export type ExtendedBoxWhere = BoxWhere & {
```
[the remark this answers is not in the document]
Comment (SepehrGanji) [reply packages/graphql-client/src/graphqlClient.ts:14]: Done!

File: packages/graphql-client/src/graphqlClient.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+      address: q.where.contract,
+      addresses: q.where.contracts,
+      ergoTreeTemplateHash: q.where.template,
+      tokenId: q.where.tokenId,
+      skip: q.skip,
+      take: q.take,
+      spent: false
+    });
+
+    // What to do about confirmed?
```
[the remark this answers is not in the document]
Comment (SepehrGanji) [reply packages/graphql-client/src/graphqlClient.ts:83]: Am I thinking complex or is this a complex query?
There is no way to include mempool in `boxes` query right now.
So I see we can:
1. Write another query, to access mempool boxes and match throughout the `txId` of them with our fetched boxes.
2. Same as 1, but matching could be done with `creationHeight`. Although I think this field is arbitrary and anyone can set their own number there.
3. Modify the `ergo-grapgql` project to support all that in one query.

For the first two ones, the matching part is easy but handling skip and limit could be tricky. Maybe I'm seeing this difficult. Please let me know what's on your mind.

File: packages/graphql-client/src/graphqlClient.ts (hunk 1 above)
Comment (SepehrGanji) [reply packages/graphql-client/src/graphqlClient.ts:116]: Okay, Done!

File: packages/graphql-client/src/graphqlClient.ts (hunk 2 above)
Comment (SepehrGanji) [reply packages/graphql-client/src/graphqlClient.ts:132]: Oh, you're right! It's optional so we can ignore it.

File: packages/graphql-client/src/graphqlClient.ts (hunk 3 above)
Comment (SepehrGanji) [reply packages/graphql-client/src/graphqlClient.ts:128]: Done!

File: packages/graphql-client/src/graphqlClient.ts [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+  /** Base16-encoded BoxIds */
+  boxIds?: HexString[];
+
+  /** Base16-encoded ErgoTrees or Base58-encoded addresses */
+  contracts?: HexString[];
+};
+
+export type GraphQLBoxQuery = BoxQuery<GraphQLBoxWhere>;
+
+export class GraphqlClient implements IChainDataClient<BoxWhere> {
```
Comment (arobsn) [inline packages/graphql-client/src/graphqlClient.ts:29]:
```suggestion
export class ErgoGraphQLClient implements IChainDataClient<BoxWhere> {
```

File: packages/graphql-client/src/graphqlClient.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (arobsn) [reply packages/graphql-client/src/graphqlClient.ts:40]: Hmm, GraphQL doesn't cast, mappings must be done manually.

File: packages/graphql-client/src/graphqlClient.ts [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+      boxId: q.where.boxId,
+      boxIds: q.where.boxIds,
+      address: q.where.contract,
+      addresses: q.where.contracts,
+      ergoTreeTemplateHash: q.where.template,
+      tokenId: q.where.tokenId,
+      skip: q.skip,
+      take: q.take,
+      spent: false
+    });
```
Comment (arobsn) [inline packages/graphql-client/src/graphqlClient.ts:86]: Let's keep query data concise by avoiding redundant query params like `boxId` and `boxIds`, you can do something like `boxIds: q.where.boxIds ?? q.where.boxId ? [q.where.boxId] : undefined`. Same for `ergoTree`, which is missing here.

File: packages/graphql-client/src/graphqlClient.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (arobsn) [reply packages/graphql-client/src/graphqlClient.ts:88]: Hmm, I realized that's a bit complex to maintain, specially with pagination. So let's ignore `includeUnconfirmed` for now and assume every returning box as `confirmed: true`.

**Review by arobsn — CHANGES_REQUESTED** [review]

File: packages/graphql-client/src/graphqlClient.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (SepehrGanji) [reply packages/graphql-client/src/graphqlClient.ts:40]: Manual mapping done!

File: packages/graphql-client/src/graphqlClient.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (SepehrGanji) [reply packages/graphql-client/src/graphqlClient.ts:88]: Done!

File: packages/graphql-client/src/graphqlClient.ts (hunk 13 above)
Comment (SepehrGanji) [reply packages/graphql-client/src/graphqlClient.ts:86]: Done!

**Comment by SepehrGanji** [comment]
@arobsn Tests added!

File: packages/graphql-client/src/utils/TestUtils.ts [hunk 14]
```diff
@@ -0,0 +1,5 @@
+export class TestUtils {
```
Comment (arobsn) [inline packages/graphql-client/src/utils/TestUtils.ts:1]: I think a class is not necessary here, you can export mockResponse directly. Also, files should be named using camelCase.

File: packages/graphql-client/src/utils/cast.ts [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+import { SignedTransaction } from "@fleet-sdk/common";
+
+export const castSignedTxToGql = (transaction: SignedTransaction): gqlSignedTransaction => {
+  return {
+    id: transaction.id,
+    inputs: transaction.inputs.map((input) => ({
+      boxId: input.boxId,
+      spendingProof: {
+        proofBytes: input.spendingProof.proofBytes,
+        extension: input.spendingProof.extension
```
Comment (arobsn) [inline packages/graphql-client/src/utils/cast.ts:11]: This mapping seems unnecessary.

**Review by arobsn — CHANGES_REQUESTED** [review]
It's looking good, left some small comments in code. 

* Please increase the code coverage, PRs need to meet 100% coverage to be merged, you can check what's not covered yet by running `pnpm cov:open`
* CI is failing for `node14`, that's because of unsupported `fetch` API. Don't worry about that, I'm considering dropping node14 support as [it's now deprecated](https://endoflife.date/nodejs)

File: packages/graphql-client/src/utils/TestUtils.ts (hunk 14 above)
Comment (SepehrGanji) [reply packages/graphql-client/src/utils/TestUtils.ts:1]: Done and done!

File: packages/graphql-client/src/utils/cast.ts (hunk 15 above)
Comment (SepehrGanji) [reply packages/graphql-client/src/utils/cast.ts:11]: Done!

**Review by arobsn — APPROVED** [review]

**Pull request merged**